### PR TITLE
API doc generation automation for dpctl Python API

### PR DESCRIPTION
**Pull request by diptorupd**
The PR incorporates various improvements to dpctl's documentation build system to automate the generation of rst files from the Python modules.

- A new `generate_rst.py` module is added that generates summary and detailed pages for dpctl Python modules and members of those modules.
- Clean ups to the CMake infrastructure to build the rst files using `generate_rst.py`.

### CODE REVIEW

**Comment by diptorupd** [comment]
TODOs

- [x] The links to in the summary tables for functions, exceptions, enums on the module summary page need to be fixed. 
- [x] I have added a small check to raise an exception whenever an undocumented object is found. The behavior will force the documentation generation to fail. I am not sure if this is a good idea or not.
- [x] The `generate_rst` script has to be integrated into the CMake build script.
- [ ] The documentation generation needs to be integrated into `setup.py`
- [x] We still need to look at the rendered paged. I have not done that locally.
- [ ] Fix docstrings in source files. This line item should be a separate PR.
- [x] Add a way to clean up the generated rst files.
- [ ] Make the `generate_rst` module independent of dpctl to enable reuse in other places.
- [x] In class summary Public and Private method heading are switched.
- [x] Check if a module has `submodules`, `enums`, `classes`, `exceptions`, `functions` before generating summary tables.
- [ ] Submodule summary pages should be included in TOC.
- [ ] Do not add `numpy_usm_shared` to docs.

**Comment by oleksandr-pavlyk** [comment]
For some reason the link to rendered pages give 404 error. Was the `index.html` moved?

**Comment by diptorupd** [comment]
b071798d560216d5eccea981af5fea586d9e3c9b broke doxyrest build.

File: docs/generate_rst.py [hunk 1]
```diff
@@ -327,15 +327,44 @@ def _generate_module_summary_rst(module):
...
+            st += stop + 1
+        # Strip out the signature in the docstring.
+        docstr = docstr[st:]
+        # The hope is to find the first line (summary) from the docstring
+        # by searching for a period followed by a new line. Not foolproof, but
+        # a best-effort check.
+        docstr = " ".join(
+            docstr[0 : docstr.find(".\n") + 1].replace("\n", " ").split()
+        )
+        if len(docstr) < 1:
```
Comment (oleksandr-pavlyk) [inline docs/generate_rst.py:365]:
```suggestion
        if not docstr:
```

Comment (diptorupd) [reply docs/generate_rst.py]: done.

**Comment by diptorupd** [comment]
@oleksandr-pavlyk Can you take a stab at the task "Do not show summary tables in autoautosummary if the specific type of member does not exist." We need to add a check in `write_rubric` in `_generate_class_rst`.

**Comment by oleksandr-pavlyk** [comment]
> @oleksandr-pavlyk Can you take a stab at the task "Do not show summary tables in autoautosummary if the specific type of member does not exist." We need to add a check in `write_rubric` in `_generate_class_rst`.

Done. I also merged the main branch into this one to prevent errors associated with `get(obj.__module__, obj.__name__)` failing for `copy_to_numpy`.

**Comment by diptorupd** [comment]
@oleksandr-pavlyk Thanks for getting the foundation laid for these changes. I am happy with these changes and they are a huge improvement over our existing docs. Especially, now that we do not have to manually keep updating the rst files. 

I have rebased to reduce noise from the commit log and am merging it to master.

**Pull request merged**